src/cascade/model/session.py

```diff
@@ -0,0 +1,552 @@
+            DismodGroups[Var]: A set of fit var.
+        """
+        if model.random_effect:
+            MATHLOG.info(f"Running fit both.")
+            return self._fit("both", model, data, initial_guess)
+        else:
+            MATHLOG.info(f"Running fit fixed.")
+            return self._fit("fixed", model, data, initial_guess)
+
+    def fit_fixed(self, model, data, initial_guess=None):
+        """Fits a model without optimizing over any random effects.
+        It does apply constant child value priors, but other random effects
+        are constrained to zero. (This is equivalent to fitting with
+        ``bound_random`` equal to zero.) This is useful when one uses fitting
+        with no random effects as a starting point for fitting with
+        random effects.
+
+        Args:
+            model (Model): A model, possibly without scale vars.
+            data (pd.DataFrame): Data to fit.
+            initial_guess (Var): Starting point to look for solutions. If not
+                given, then the mean of the priors is taken as the initial
+                guess.
+
+        Returns:
+            DismodGroups[Var]: A set of fit var.
+        """
+        return self._fit("fixed", model, data, initial_guess)
+
+    def fit_random(self, model, data, initial_guess=None):
+        """
+        Fits the data with the model.
+        This optimizes the random effects with the fixed effects set to their
+        starting values. The fixed effects are unchanged.
+
+        Args:
+            model (Model): A model, possibly without scale vars.
+            data (pd.DataFrame): Data to fit.
+            initial_guess (Var): Starting point to look for solutions. If not
+                given, then the mean of the priors is taken as the initial
+                guess.
+
+        Returns:
+            DismodGroups[Var]: A set of fit var.
+        """
+        return self._fit("random", model, data, initial_guess)
+
+    def _fit(self, fit_level, model, data, initial_guess):
+        self._setup_model(model, data)
+        if initial_guess is not None:
+            MATHLOG.info(f"Setting initial value for search from user argument.")
+            self.set_var("start", initial_guess)
+        # else use the one generated by the call to init, coming from the mean.
+        self._run_dismod(["fit", fit_level])
+        return FitResult(self, self.get_var("fit"))
+
+    def _setup_model(self, model, data):
+        extremal = list()
+        if data is not None and not data.empty:
+            for dimension in ["age", "time"]:
+                cols = [ac for ac in data.columns if ac.startswith(dimension)]
+                if not cols:
+                    raise ValueError(f"Dataframe must have age and time columns but has {data.columns}.")
+                extremal.append({data[cols].min().min(), data[cols].max().max()})
+        self.write_model(model, extremal)
+        self.write_data(data)
+        self._run_dismod(["init"])
```

Comment:
It sometimes convenient to be able to setup the database without actually invoking dismod at all. For debugging. I'm not sure we want to support that and maybe it's less important now that things are more mature.

Reply:
The current system has the --db-only flag. Though that may still run init. I can't remember. You could pass a similar flag here, though that's kind of ugly.

Reply:
I'd rather put an option on the session to a) keep a copy of the db before each run of a Dismod-AT command or b) never run Dismod-AT and mock all return values.

Reply:
Let's call that a new feature. Not on this PR.

Reply:
Yep